clang/lib/Gold/GoldDeclarationBuilder.cpp

```diff
@@ -0,0 +1,814 @@
+  //   clang::DeclarationNameInfo DNI({TheDecl->getId()}, TheDecl->IdDcl->getLoc());
+  //   clang::LookupResult R(SemaRef.getCxxSema(), DNI, clang::Sema::LookupAnyName);
+  //   if (SemaRef.lookupUnqualifiedName(R, SemaRef.getCurrentScope())) {
+  //     return true;
+  //   }
+  // }
+  // Doing simple blook lookup.
+  clang::IdentifierInfo* Id = TheDecl->getId();
+  Scope *CurScope = SemaRef.getCurrentScope();
+  if (CurScope->isBlockScope()) {
+    // If we're assigning to a name that already exist in the current block,
+    // then we're not declaring anything. For example:
+    // \code
+    //    x = 3
+    //    x = 4
+    // \endcode
+    // The first statement is a declaration. The second is an assignment.
+    // FIXME: is this the right way to handle the lookup set?
+    if (OperatorEquals && !CurScope->findDecl(Id).empty())
+      return true;
+  }
+
+  if (IsInsideEnum) {
+    return checkEnumDeclaration(DeclExpr, TheDecl);
+  }
+  return false;
+}
+
+bool DeclarationBuilder::checkEnumDeclaration(const Syntax *DeclExpr,
+                                              Declaration *TheDecl) {
+  // We know we are 100% have to be this or an error.
+  TheDecl->SuspectedKind = UDK_EnumConstant;
+  // Because we are inside of an enum we are 100% sure that this is an error.
+  if (TheDecl->GlobalNsSpecifier) {
+    SemaRef.Diags.Report(TheDecl->FunctionDcl->getLoc(),
+                         clang::diag::err_invalid_declarator_sequence)
+                         << 4;
+    return true;
+  }
+
+  if (!TheDecl->NNSInfo.empty()) {
+    SemaRef.Diags.Report(TheDecl->FunctionDcl->getLoc(),
+                         clang::diag::err_invalid_declarator_sequence)
+                         << 9;
+    return true;
+  }
+
+  if (TheDecl->FunctionDcl) {
+    SemaRef.Diags.Report(TheDecl->FunctionDcl->getLoc(),
+                         clang::diag::err_invalid_declarator_sequence)
+                         << 0;
+    return true;
+  }
+
+  if (TheDecl->TypeDcl) {
+    SemaRef.Diags.Report(TheDecl->FunctionDcl->getLoc(),
+                         clang::diag::err_invalid_declarator_sequence)
+                         << 9;
+    return true;
+  }
+
+  if (TheDecl->TemplateParameters) {
+    SemaRef.Diags.Report(TheDecl->FunctionDcl->getLoc(),
+                         clang::diag::err_invalid_declarator_sequence)
+                         << 2;
+    return true;
+  }
+
+  if (TheDecl->SpecializationArgs) {
+    SemaRef.Diags.Report(TheDecl->FunctionDcl->getLoc(),
+                         clang::diag::err_invalid_declarator_sequence)
+                         << 8;
+    return true;
+  }
+
+  // TODO: It may be necessary to specifically include a test for
+  // TheDecl->Init
+  if (TheDecl->Op) {
+    if (const auto *Call = dyn_cast<CallSyntax>(TheDecl->Op)) {
+      if (const auto *Operator = dyn_cast<AtomSyntax>(Call->getCallee())) {
+        if (Operator->getSpelling() != "operator'='") {
```

Comment:
Try to avoid string comparisons like this, we have a tokenization of almost every kind of operator in Sema at this point.

`getFusedOpKind` is always preferred.